go/worker/sentry/grpc/worker.go

```diff
@@ -0,0 +1,199 @@
+	// Per service polcy checkers.
+	grpcPolicyCheckers map[cmnGrpc.ServiceName]*policy.DynamicRuntimePolicyChecker
+
+	*upstreamConn
+
+	grpc     *cmnGrpc.Server
+	identity *identity.Identity
+}
+
+type upstreamConn struct {
+	conn              *grpc.ClientConn
+	resolverCleanupCb func()
+}
+
+// TODO: this whole authFunction is currently hardcoded to only supports 'storage' backend,
+//       so that i was able to do test. Sentry GRPC worker should be upstream agnostic,
+//       but need to do some more policy refacotrings frits.
+func (g *Worker) authFunction() auth.DefaultAuthFunc {
+	return func(ctx context.Context,
+		fullMethodName string,
+		req interface{}) error {
+
+		g.RLock()
+		defer g.RUnlock()
+		p, ok := g.grpcPolicyCheckers[storageAPI.ServiceName]
+		if !ok {
+			// TODO: this means we need to refactor policies a bit so that
+			// those are defined for every method.
+			return status.Errorf(codes.Unauthenticated, "no policy defined on endpoint")
+		}
+
+		// TODO: refactor polciies so that this doesn't need to be here, but instead a single "CheckAccessAllowed"
+		// is enough to handle all cases & multiple upstream services.
+		switch r := req.(type) {
+		case *storageAPI.ApplyRequest:
+			return p.CheckAccessAllowed(ctx, accessctl.Action(fullMethodName), r.Namespace)
+		case *storageAPI.ApplyBatchRequest:
+			return p.CheckAccessAllowed(ctx, accessctl.Action(fullMethodName), r.Namespace)
+		case *storageAPI.MergeRequest:
+			return p.CheckAccessAllowed(ctx, accessctl.Action(fullMethodName), r.Namespace)
+		case *storageAPI.MergeBatchRequest:
+			return p.CheckAccessAllowed(ctx, accessctl.Action(fullMethodName), r.Namespace)
+		case *storageAPI.GetDiffRequest:
+			return p.CheckAccessAllowed(ctx, accessctl.Action(fullMethodName), r.StartRoot.Namespace)
+		case *storageAPI.GetCheckpointRequest:
+			return p.CheckAccessAllowed(ctx, accessctl.Action(fullMethodName), r.Root.Namespace)
+		default:
+			return nil
+		}
+	}
+}
+
+func (g *Worker) updatePolicies(p policyAPI.ServicePolicies) {
+	g.logger.Debug("updating policies",
+		"policy", p,
+	)
+
+	g.Lock()
+	defer g.Unlock()
+	g.grpcPolicyCheckers[p.Service] = policy.NewDynamicRuntimePolicyChecker(p.Service, nil)
+	for namespace, policy := range p.AccessPolicies {
+		g.grpcPolicyCheckers[p.Service].SetAccessPolicy(policy, namespace)
+	}
+
+}
+
+func (g *Worker) worker() {
+	go func() {
+		defer close(g.quitCh)
+		defer (g.cancelCtx)()
+		<-g.grpc.Quit()
+	}()
+
+	g.policyWatcher = policyAPI.NewPolicyWatcherClient(g.conn)
+
+	// TODO: should close this after getting intial policy.
+	close(g.initCh)
+
+	pCh, pSub, err := g.policyWatcher.WatchPolicies(g.ctx)
+	if err != nil {
+		g.logger.Error("failed to watch policies",
+			"err", err,
+		)
+		panic("failed to watch policies")
+	}
+	defer pSub.Close()
+
+	// Watch Policies.
+	for {
+		select {
+		case p, ok := <-pCh:
+			if !ok {
+				g.logger.Error("WatchPolicies stream failure")
+				// TODO: Panic? Retry?
```

Comment:
When can this happen?

Reply:
i think connection errors with the upstream (for whatever reason) would make this fail